### PR TITLE
Fixed JsonDataContract.GetGeneratedReadWriteDelegates.

### LINKED ISSUE #12176: JsonDataContract.GetGeneratedReadWriteDelegates May Throw ArgumentNullException
JsonDataContract.GetGeneratedReadWriteDelegates may throw ArgumentNullException in NetNative.

```
--------------------------------
Running Test: DataContractJsonSerializerTests.DCJS_ValidateExceptionOnUnspecifiedRootSerializationType
Caught Unexpected exception:System.ArgumentNullException: ArgumentNull_Generic
Arg_ParamName_Name
   at System.Collections.Generic.Dictionary$2<System.__Canon,System.__Canon>.FindEntry(__Canon key) in f:\dd\ndp\fxcore\src\
System.Collections\System\Collections\Generic\Dictionary.cs:line 266
   at System.Collections.Generic.Dictionary$2<System.__Canon,System.__Canon>.TryGetValue(__Canon key, __Canon& value) in f:\
dd\ndp\fxcore\src\System.Collections\System\Collections\Generic\Dictionary.cs:line 434
   at System.Runtime.Serialization.Json.JsonDataContract.GetGeneratedReadWriteDelegates($DataContract c)
   at System.Runtime.Serialization.Json.JsonDataContract.TryGetReadWriteDelegatesFromGeneratedAssembly($DataContract c)
   at System.Runtime.Serialization.Json.JsonClassDataContract.get_JsonFormatWriterDelegate()
   at System.Runtime.Serialization.Json.JsonClassDataContract.WriteJsonValueCore($XmlWriterDelegator jsonWriter, Object obj,
 $XmlObjectSerializerWriteContextComplexJson context, RuntimeTypeHandle declaredTypeHandle)
   at System.Runtime.Serialization.Json.JsonDataContract.WriteJsonValue($XmlWriterDelegator jsonWriter, Object obj, $XmlObje
ctSerializerWriteContextComplexJson context, RuntimeTypeHandle declaredTypeHandle)
   at System.Runtime.Serialization.Json.DataContractJsonSerializerImpl.WriteJsonValue($JsonDataContract contract, $XmlWriter
Delegator writer, Object graph, $XmlObjectSerializerWriteContextComplexJson context, RuntimeTypeHandle declaredTypeHandle)
   at System.Runtime.Serialization.Json.XmlObjectSerializerWriteContextComplexJson.WriteDataContractValue($DataContract data
Contract, $XmlWriterDelegator xmlWriter, Object obj, RuntimeTypeHandle declaredTypeHandle)
   at System.Runtime.Serialization.XmlObjectSerializerWriteContext.SerializeWithoutXsiType($DataContract dataContract, $XmlW
riterDelegator xmlWriter, Object obj, RuntimeTypeHandle declaredTypeHandle)
   at System.Runtime.Serialization.Json.DataContractJsonSerializerImpl.InternalWriteObjectContent($XmlWriterDelegator writer
, Object graph)
   at System.Runtime.Serialization.Json.DataContractJsonSerializerImpl.InternalWriteObject($XmlWriterDelegator writer, Objec
t graph)
   at System.Runtime.Serialization.XmlObjectSerializer.InternalWriteObject($XmlWriterDelegator writer, Object graph, $DataCo
ntractResolver dataContractResolver)
   at System.Runtime.Serialization.XmlObjectSerializer.WriteObjectHandleExceptions($XmlWriterDelegator writer, Object graph,
 $DataContractResolver dataContractResolver)
   at System.Runtime.Serialization.XmlObjectSerializer.WriteObjectHandleExceptions($XmlWriterDelegator writer, Object graph)

   at System.Runtime.Serialization.Json.DataContractJsonSerializerImpl.WriteObject($XmlDictionaryWriter writer, Object graph
)
   at System.Runtime.Serialization.Json.DataContractJsonSerializerImpl.WriteObject(Stream stream, Object graph)
   at System.Runtime.Serialization.Json.DataContractJsonSerializer.WriteObject(Stream stream, Object graph)
   at DataContractJsonSerializerTests.DCJS_ValidateExceptionOnUnspecifiedRootSerializationType() in d:\dd\ProjectN\src\QA\To
F\tests\FX\Conformance\System.Runtime.Serialization.Json\4.0.0.0\DataContractJsonSerializer.NetNative.cs:line 36
   at EntryPointMain.<Main>b__69() in d:\dd\ProjectN\src\QA\ToF\PN\x86\dbg\IL\FX\Conformance\System.Runtime.Serialization.Js
on\4.0.0.0\DataContractJsonSerializerTests.main.cs:line 118
   at System.Action.InvokeOpenStaticThunk()
   at System.Action.Invoke()
   at CoreFXTestLibrary.Internal.Runner.RunTestMethod() in d:\dd\ProjectN\src\QA\ToF\tests\FX\Conformance\Common\Xunit\Inter
nal\Runner.cs:line 170
   at CoreFXTestLibrary.Internal.Runner.RunTest($TestInfo t) in d:\dd\ProjectN\src\QA\ToF\tests\FX\Conformance\Common\Xunit\
Internal\Runner.cs:line 95
---- Test FAILED ---------------
```

### PR DESCRIPTION
**Pull request by shmao**
Fixed the issue that JsonDataContract.GetGeneratedReadWriteDelegates may throw ArgumentNullException in NetNative.

Fix #12176

/cc: @mconnew @huanwu @zhenlan

### CODE REVIEW

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/JsonDataContract.cs [hunk 1]
```diff
@@ -63,8 +63,10 @@ public static JsonReadWriteDelegates GetGeneratedReadWriteDelegates(DataContract
 #if NET_NATIVE
             // The c passed in could be a clone which is different from the original key,
             // We'll need to get the original key data contract from generated assembly.
-            DataContract keyDc = DataContract.GetDataContractFromGeneratedAssembly(c.UnderlyingType);
-            return JsonReadWriteDelegates.GetJsonDelegates().TryGetValue(keyDc, out result) ? result : null;
+            DataContract keyDc = (c != null && c.UnderlyingType != null) ?
```
Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/JsonDataContract.cs]: Changed per the suggestion. Thanks!

**Comment by mconnew** [comment]
:shipit:

**Comment by shmao** [comment]
@dotnet-bot test Innerloop Ubuntu14.04 Debug Build and Test

**Comment by shmao** [comment]
Innerloop Ubuntu14.04 Debug Build and Test failed for the following error,

```
/mnt/j/workspace/dotnet_corefx/master/ubuntu14.04_debug_prtest/Tools/VersionTools.targets(47,5): error : Dependencies invalid: In '/mnt/j/workspace/dotnet_corefx/master/ubuntu14.04_debug_prtest/src/System.Net.Sockets/ref/project.json', 'System.IO 4.3.0-beta-24604-02' must be '4.4.0-beta-24607-03' (CoreFx)
```

**Comment by shmao** [comment]
@stephentoub , the error above doesn't seem related to the fix. Can we merge this PR? Thanks!

**Comment by shmao** [comment]
@dotnet-bot test Innerloop Ubuntu14.04 Debug Build and Test

**Pull request merged**